sdk/python/feast/infra/online_stores/contrib/hbase_online_store/hbase.py

```diff
@@ -0,0 +1,234 @@
+
+    @property
+    def real_conn(self) -> Connection:
+        """Stores the real happybase Connection to connect to hbase."""
+        return self._real_conn
+
+    def close(self) -> None:
+        """Close the happybase connection."""
+        self.real_conn.close()
+
+
+class HbaseOnlineStore(OnlineStore):
+    """
+    Online feature store for Hbase.
+
+    Attributes:
+        _conn: Happybase Connection to connect to hbase thrift server.
+    """
+
+    _conn: Connection = None
+
+    def _get_conn(self, config: RepoConfig):
+        """
+        Get or Create Hbase Connection from Repoconfig.
+
+        Args:
+            config: The RepoConfig for the current FeatureStore.
+        """
+
+        store_config = config.online_store
+        assert isinstance(store_config, HbaseOnlineStoreConfig)
+
+        if not self._conn:
+            self._conn = Connection(host=store_config.host, port=int(store_config.port))
+        return self._conn
+
+    def online_write_batch(
+        self,
+        config: RepoConfig,
+        table: FeatureView,
+        data: List[
+            Tuple[EntityKeyProto, Dict[str, ValueProto], datetime, Optional[datetime]]
+        ],
+        progress: Optional[Callable[[int], Any]],
+    ) -> None:
+        """
+        Write a batch of feature rows to Hbase online store.
+
+        Args:
+            config: The RepoConfig for the current FeatureStore.
+            table: Feast FeatureView.
+            data: a list of quadruplets containing Feature data. Each quadruplet contains an Entity Key,
+            a dict containing feature values, an event timestamp for the row, and
+            the created timestamp for the row if it exists.
+            progress: Optional function to be called once every mini-batch of rows is written to
+            the online store. Can be used to display progress.
+        """
+
+        hbase = HbaseUtils(self._get_conn(config))
+        project = config.project
+        table_name = _table_id(project, table)
+
+        b = hbase.batch(table_name)
+        for entity_key, values, timestamp, created_ts in data:
+            row_key = serialize_entity_key(entity_key).hex()
+            values_dict = {}
+            for feature_name, val in values.items():
+                values_dict[
+                    HbaseConstants.get_col_from_feature(feature_name)
+                ] = val.SerializeToString()
+            if isinstance(timestamp, datetime):
+                values_dict[HbaseConstants.DEFAULT_EVENT_TS] = struct.pack(
+                    ">L", int(calendar.timegm(timestamp.timetuple()))
+                )
+            else:
+                values_dict[HbaseConstants.DEFAULT_EVENT_TS] = timestamp
+            if created_ts is not None:
+                if isinstance(created_ts, datetime):
+                    values_dict[HbaseConstants.DEFAULT_CREATED_TS] = struct.pack(
+                        ">L", int(calendar.timegm(created_ts.timetuple()))
+                    )
```

Comment:
Does this field have some specific significance/format for HBase? And `calendar` is the canonical way to generate this value?

Reply:
**Does this field have some specific significance/format for HBase?** 

In Hbase, every data is stored as byte array. To store datetime columns in HBase, I choose to convert the byte array to time since epoch and then convert the integer to byte array. 

**`calendar` is the canonical way to generate this value?**

There are a couple of ways to convert timestamp to epoch. 

- calendar.timegm()
- time.mktime()

[time.mktime()](http://docs.python.org/library/time.html#time.mktime) assumes that the passed tuple is in local time, [calendar.timegm()](http://docs.python.org/library/calendar.html#calendar.timegm) assumes it's in GMT/UTC. Depending on the interpretation the tuple represents a different time, so the functions return different values (seconds since the epoch are UTC based)